### PR TITLE
HDDS-5370. [FSO] Handle OMClientRequest based on the bucket layout.

### PR DESCRIPTION
**Pull request by aryangupta1998**
## What changes were proposed in this pull request?

Now we determine if the bucket is FSOptimized if the config 'ozone.om.enable.filesystem.paths' is set to true. In this and follow-ups jiras we will remove this property and determine the bucket layout by directly comparing it with an enum(BucketType.FILE_SYSTEM_OPTIMIZED).

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5370

## How was this patch tested?

Tested Manually.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java [hunk 1]
```diff
@@ -437,4 +447,117 @@ private static ServiceException createLeaderNotReadyException(
...
+      volName = omRequest.getCompleteMultiPartUploadRequest().getKeyArgs()
+          .getVolumeName();
+      buckName = omRequest.getCompleteMultiPartUploadRequest().getKeyArgs()
+          .getBucketName();
+      buckInfo = getOmBucketInfo(ozoneManager, buckInfo, volName, buckName);
+      bucketLayout = getBucketLayout(buckInfo);
+      break;
+    default:
+      break;
+    }
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:536]: `bucketLayout.equals(BucketLayout.FILE_SYSTEM_OPTIMIZED)`, there is a findbug warn, please modify like below.

bucketLayout can be null, please do `BucketLayout.FILE_SYSTEM_OPTIMIZED.equals(bucketLayout)`.

Also, please add a TODO to handle the case of null buckInfo.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java [hunk 2]
```diff
@@ -281,10 +287,12 @@ private static OMClientRequest getOMAclRequest(OMRequest omRequest) {
       } else if (ObjectType.BUCKET == type) {
         return new OMBucketRemoveAclRequest(omRequest);
       } else if (ObjectType.KEY == type) {
-        if (isBucketFSOptimized()){
+        OMKeyRemoveAclRequest aclReq = new OMKeyRemoveAclRequest(omRequest);
+        OmBucketInfo buckInfo = aclReq.getBucketInfo(ozoneManager);
+        if (buckInfo.equals(BucketLayout.FILE_SYSTEM_OPTIMIZED)) {
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:292]: `buckInfo.equals(BucketLayout.FILE_SYSTEM_OPTIMIZED)`
There is a findbug on this statement.

Please use getLayout while comparing, spomething like `buckInfo.getLaytout().equals(BucketLayout.FILE_SYSTEM_OPTIMIZED)`

**Review by rakeshadr — COMMENTED** [review]
Thanks @aryangupta1998 for the patch. Added a few comments and will do a detailed review later.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java (hunk 1 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:537]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java (hunk 2 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java]: Done

**Comment by rakeshadr** [comment]
@aryangupta1998 There are test case failures, due to the refactored logic of isBucketFSOptimized. Please skip the `isBucketFSOptimized` check by setting the flag to `isFSOBucketCheckEnabled` false.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java [hunk 3]
```diff
@@ -3093,4 +3095,25 @@ public OmKeyInfo getPendingDeletionDir() throws IOException {
 
     return files;
   }
+
+  public boolean isBucketFSOptimized(String volName, String buckName)
+      throws IOException {
+    if (!isFSOBucketCheckEnabled) {
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:3101]: On a second thought, like we discussed offline we can add a safer null check for the `ozoneManager == null` instead of using the `isFSOBucketCheckEnabled` test flag. ozoneManager can be null for non-FSO test cases and will execute non-FSO code path. 
But, please ensure that all `*FSO test cases` would see a non-null ozoneManager to cover the FSO execution code path.

https://github.com/apache/ozone/pull/2533/files#diff-bde0dade7dd5ddda419499f4f999d25d40fcec1412e0ce809c36ffd1be473f22R3101


  ```
public boolean isBucketFSOptimized(String volName, String buckName)
      throws IOException {
    // safer check for unit test case. In reality ozoneManager will never be null
    if (ozoneManager == null) {
      return false;
    } 
```

Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyDeleteRequest.java [hunk 4]
```diff
@@ -197,4 +198,118 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
     return omClientResponse;
   }
 
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyDeleteRequest.java:201]: Can we avoid copying the same method here? We can call the new method with null bucket layout in the above method.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestOMRequestUtils.java [hunk 5]
```diff
@@ -439,6 +441,21 @@ public static void addBucketToDB(String volumeName, String bucketName,
         new CacheValue<>(Optional.of(omBucketInfo), 1L));
   }
 
+  public static void addBucketToDB(String volumeName, String bucketName,
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestOMRequestUtils.java:444]: same here. Can we avoid copying the method?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyDeleteRequest.java (hunk 4 above)
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyDeleteRequest.java:201]: Why do we need bucketLayout for keyDelete? We deduce the bucketLayout in OzoneManagerRatisUtils and create appropriate request right . Is this method only created for unit test?

**Review by sadanand48 — COMMENTED** [review]
Thanks @aryangupta1998 for the patch. Added few comments.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java [hunk 6]
```diff
@@ -279,6 +279,7 @@ protected OmMetadataManagerImpl() {
 
   @Override
   public Table<String, OmKeyInfo> getKeyTable() {
+    // TODO: Refactor the below function by reading bucketLayout.
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java:282]: Please raise a jira task to address this TODO for tracking it. Also, please mention the jira id here in TODO section.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java [hunk 7]
```diff
@@ -437,4 +450,117 @@ private static ServiceException createLeaderNotReadyException(
     return new ServiceException(leaderNotReadyException);
   }
 
+  public static boolean isBucketFSOptimized(OMRequest omRequest,
+      OzoneManager ozoneManager) {
+    Type cmdType = omRequest.getCmdType();
+    BucketLayout bucketLayout = null;
+    OmBucketInfo buckInfo = null;
+    switch (cmdType) {
+    case AllocateBlock:
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:459]: How about avoiding these duplicate switch statement using following approach:

- Add new function which takes args like below:

```
  private static boolean isBucketFSOptimized(String volName, String buckName,
      OzoneManager ozoneManager) {
    BucketLayout bucketLayout = null;
    OmBucketInfo buckInfo = null;
    buckInfo = getOmBucketInfo(ozoneManager, buckInfo, volName, buckName);
    bucketLayout = getBucketLayout(buckInfo);
    return BucketLayout.FILE_SYSTEM_OPTIMIZED.equals(bucketLayout);
  }
```
- Call the above function from the existing switch-case like below. The same can be repeated to other bucket layout cases as well : CreateKey, CommitKey etc 

```
    case AllocateBlock:
      keyArgs = omRequest.getAllocateBlockRequest().getKeyArgs();
      if (isBucketFSOptimized(keyArgs.getVolumeName(), keyArgs.getBucketName(),
          ozoneManager)) {
        return new OMAllocateBlockRequestWithFSO(omRequest);
      }
      return new OMAllocateBlockRequest(omRequest);

    case CreateKey:
      keyArgs = omRequest.getCreateKeyRequest().getKeyArgs();
      if (isBucketFSOptimized(keyArgs.getVolumeName(), keyArgs.getBucketName(),
          ozoneManager)) {
        return new OMKeyCreateRequestWithFSO(omRequest);
      }
      return new OMKeyCreateRequest(omRequest);

.....
....
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyDeleteRequest.java (hunk 4 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyDeleteRequest.java:201]: Done

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestOMRequestUtils.java (hunk 5 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestOMRequestUtils.java:444]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyDeleteRequest.java (hunk 4 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyDeleteRequest.java:201]: Yes, this method is necessary for the test. Also, we need `bucketLayout` as a parameter because we need to set the `bucketLayout` in the `OmBucketInfo`, earlier for FSO(File System Optimized) cases we were dependent on a flag `isBucketFSOptimized` but with the help of this PR, we will deduce the `bucketLayout` directly from `OmBucketInfo` hence, we need to set the appropriate `bucketLayout`. I hope this clears your doubt, if you have more questions please feel free to add them!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java [hunk 8]
```diff
@@ -437,4 +450,117 @@ private static ServiceException createLeaderNotReadyException(
...
+      String volName =
+          omRequest.getAllocateBlockRequest().getKeyArgs().getVolumeName();
+      String buckName =
+          omRequest.getAllocateBlockRequest().getKeyArgs().getBucketName();
+      buckInfo = getOmBucketInfo(ozoneManager, buckInfo, volName, buckName);
```
[the remark this answers is not in the document]
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:464]: I have made the necessary changes according to your comment below.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java (hunk 6 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java:282]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java (hunk 7 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:459]: Thanks, @rakeshadr for this nice approach, I have made the necessary changes.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java [hunk 9]
```diff
@@ -437,4 +473,36 @@ private static ServiceException createLeaderNotReadyException(
...
 
+  private static OmBucketInfo getOmBucketInfo(OzoneManager ozoneManager,
+      OmBucketInfo buckInfo, String volName, String buckName) {
+    String buckKey =
+        ozoneManager.getMetadataManager().getBucketKey(volName, buckName);
+    try {
+      buckInfo =
+          ozoneManager.getMetadataManager().getBucketTable().get(buckKey);
+    } catch (IOException e) {
+      e.printStackTrace();
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:484]: @aryangupta1998 
Please add log message and remove e.printStackTrace() statement.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java [hunk 10]
```diff
@@ -146,6 +148,30 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
...
+      String bucket = objectParser.getBucket();
+
+      String buckKey =
+          ozoneManager.getMetadataManager().getBucketKey(volume, bucket);
+
+      try {
+        buckInfo =
+            ozoneManager.getMetadataManager().getBucketTable().get(buckKey);
+      } catch (IOException e) {
+        e.printStackTrace();
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java:167]: Please add log message and remove e.printStackTrace() statement.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java [hunk 11]
```diff
@@ -437,4 +473,36 @@ private static ServiceException createLeaderNotReadyException(
...
+    }
+    return buckInfo;
+  }
+
+  private static BucketLayout getBucketLayout(OmBucketInfo buckInfo) {
+    if (buckInfo != null) {
+      return buckInfo.getBucketLayout();
+    } else {
+      buckInfo = null;
+      // TODO: Handle bucket validation
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:494]: bucket null is a tricky case. Since this bucketTable is not inside BUCKET_LOCK the consistency won't be guaranteed. I think, if the bucket is coming as null then how about acquiring a BUCK_LOCK and re-read the bucket key again?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java [hunk 12]
```diff
@@ -437,4 +450,117 @@ private static ServiceException createLeaderNotReadyException(
...
+      break;
+    default:
+      break;
+    }
+    return BucketLayout.FILE_SYSTEM_OPTIMIZED.equals(bucketLayout);
+  }
+
+  private static OmBucketInfo getOmBucketInfo(OzoneManager ozoneManager,
+      OmBucketInfo buckInfo, String volName, String buckName) {
+    String buckKey =
```
[the remark this answers is not in the document]
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:478]: Thanks @bharatviswa504 for the comment. Now, with this patch`#getBucketTable().get(buckKey)` is coming in the hot path for all the key requests, we thought of avoiding acquiring the costlier BUCKET_LOCK. As `BucketInfo#layout` is an immutable entity, I think it won't affect correctness. Does it make sense to you?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java (hunk 9 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java (hunk 10 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:486]: Thanks @bharatviswa504 for the offline discussion.

@aryangupta1998 Here all the client requests are executed in sequential model, so not required to acquire a BUCKET_LOCK. All the client requests are executed through `OzoneManagerStateMachine#runCommand` function and ensures sequential execution path.

Below is the call trace to perform OM client request operation:

`OzoneManagerStateMachine#applyTransaction -> OzoneManagerStateMachine#runCommand -> OzoneManagerRequestHandler#handleWriteRequest -> OzoneManagerRatisUtils#createClientRequest -> OzoneManagerRatisUtils#getOmBucketInfo -> omMetadataManager().getBucketTable().get(buckKey)`

@aryangupta1998 could you please add java comment about this part.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java [hunk 13]
```diff
@@ -437,4 +450,117 @@ private static ServiceException createLeaderNotReadyException(
...
+        ozoneManager.getMetadataManager().getBucketKey(volName, buckName);
+    try {
+      buckInfo =
+          ozoneManager.getMetadataManager().getBucketTable().get(buckKey);
+    } catch (IOException e) {
+      e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java]: Done!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java (hunk 11 above)
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:502]: @aryangupta1998 Since `#validateBucketAndVolume()` check is there in every OMClientRequest , how about to return `BucketLayout.LEGACY` in case of a null bucket. For debugging purpose, please add warn log message saying that bucket doesn't exists.

`LOG.error("Bucket not found: {}/{} ", volumeName, bucketName);`

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java [hunk 14]
```diff
@@ -267,10 +302,13 @@ private static OMClientRequest getOMAclRequest(OMRequest omRequest) {
       } else if (ObjectType.BUCKET == type) {
         return new OMBucketAddAclRequest(omRequest);
       } else if (ObjectType.KEY == type) {
-        if (isBucketFSOptimized()){
+        OMKeyAddAclRequest aclReq = new OMKeyAddAclRequest(omRequest);
+        OmBucketInfo buckInfo = aclReq.getBucketInfo(ozoneManager);
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:306]: @aryangupta1998  Like we discussed for other client KeyRequests, `bucketInfo` can be null. [Reference comment](https://github.com/apache/ozone/pull/2533/files#r692388618)

Instead of returning BucketInfo, you can return BucketLayout.
Just rename method `OMKeyAclRequest.getBucketInfo` to `OMKeyAclRequestgetBucketLayout` and then return Layout itself. Here, you can handle bucketInfo null check and return BucketLayout.LEGACY.

For debugging purpose, please add warn log message saying that bucket doesn't exists.
LOG.error("Bucket not found: {}/{} ", volumeName, bucketName);

Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:499]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java (hunk 11 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:516]: Done

Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:516]: Thanks @aryangupta1998 for addressing this. I didn't think about a unit test case verifying the default `BucketLayout.LEGACY` path in my previous comment. Can we add a new test case where the bucket doesn't exists and verify it. Probably you can add a simple test in TestRootedOzoneFileSystem or TestRootedOzoneFileSystemWithFSO.

Verify creating a file or key on a non-existent bucket something like,

`bucketPath = new Path(volumePath, "invalidBucket");`

**Comment by sadanand48** [comment]
+1. LGTM

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java (hunk 11 above)
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:516]: @aryangupta1998 Please refer below unit test case and can use this kinda test to verify our code path. Since all the RPC key calls are invoked on bucket, a mock based unit test will work.
```
  @Test(timeout = 300_000)
  public void testRequestWithNonExistentBucket() throws Exception {
    ozoneManager = Mockito.mock(OzoneManager.class);
    ozoneConfiguration.set(OMConfigKeys.OZONE_OM_DB_DIRS,
        folder.newFolder().getAbsolutePath());
    omMetadataManager = new OmMetadataManagerImpl(ozoneConfiguration);
    when(ozoneManager.getMetadataManager()).thenReturn(omMetadataManager);

    String volumeName = "vol1";
    String bucketName = "invalidBuck";
    OzoneManagerProtocolProtos.OMRequest omRequest =
        TestOMRequestUtils.createCompleteMPURequest(volumeName, bucketName,
            "mpuKey", "mpuKeyID", new ArrayList<>());

    OMClientRequest req = OzoneManagerRatisUtils.createClientRequest(omRequest, ozoneManager);
    Assert.assertNotNull(req);
    Assert.assertTrue("Unexpected request on invalid bucket", req instanceof S3MultipartUploadCompleteRequest);
  }
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java [hunk 15]
```diff
@@ -146,6 +153,36 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
...
+        if (buckInfo == null) {
+          LOG.error("Bucket not found: {}/{} ", volume, bucket);
+          return BucketLayout.LEGACY;
+        }
+        bucketLayout = buckInfo.getBucketLayout();
+      } catch (IOException e) {
+        LOG.debug("Failed to get the value for the key: " + buckKey);
+      }
+    } catch (OMException ome) {
+      // Handle exception
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java:181]: Please add error log message here. Its not required to throw exception explicitly and the current handling of returning default `BucketLayout.LEGACY` type is OK.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java [hunk 16]
```diff
@@ -146,6 +153,36 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
...
+      try {
+        buckInfo =
+            ozoneManager.getMetadataManager().getBucketTable().get(buckKey);
+        if (buckInfo == null) {
+          LOG.error("Bucket not found: {}/{} ", volume, bucket);
+          return BucketLayout.LEGACY;
+        }
+        bucketLayout = buckInfo.getBucketLayout();
+      } catch (IOException e) {
+        LOG.debug("Failed to get the value for the key: " + buckKey);
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java:178]: Please change it to` LOG.error` 

`LOG.error("Failed to get bucket for the key: {}", buckKey);`

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java [hunk 17]
```diff
@@ -240,11 +240,17 @@ public OMResponse handleReadRequest(OMRequest request) {
...
+    OMClientRequest omClientRequest = null;
+    OMClientResponse omClientResponse = null;
+    try {
+      omClientRequest =
+          OzoneManagerRatisUtils.createClientRequest(omRequest, impl);
+      omClientResponse = omClientRequest
+          .validateAndUpdateCache(getOzoneManager(), transactionLogIndex,
+              ozoneManagerDoubleBuffer::add);
+    } catch (IOException e) {
+      LOG.debug("Exception: " + e);
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:252]: Since we are not throwing `OzoneManagerRatisUtils.createClientRequest` exception, we can revert this change, isn't it?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java [hunk 18]
```diff
@@ -140,8 +142,18 @@ public static void setBucketFSOptimized(boolean enabledFSO) {
    * @return OMClientRequest
    * @throws IOException
    */
-  public static OMClientRequest createClientRequest(OMRequest omRequest) {
+  public static OMClientRequest createClientRequest(OMRequest omRequest,
+      OzoneManager ozoneManager) throws IOException {
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:146]: Please remove `throws IOException` from the method signature, which is not required as we are handling exception using the BucketLayout.LEGACY, right?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java (hunk 11 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:516]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java (hunk 15 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java:181]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java (hunk 16 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAclRequest.java:178]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java (hunk 17 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java:252]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java (hunk 18 above)
Comment (aryangupta1998) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/utils/OzoneManagerRatisUtils.java:146]: Done

**Review by rakeshadr — APPROVED** [review]
+1 LGTM, I will merge it shortly.

Thanks @aryangupta1998 for the continuous efforts.
Thanks @bharatviswa504 , @sadanand48 for the useful reviews.
